File: problem_6/sentimental-readability/readability.py
from cs50 import get_string

# get string from user
text = None
while True:
    try:
        text = get_string('Text: ')
        break
    except ValueError:
        ...

# counting words
def count_word(text):
    words = text.count(' ')
    return words + 1

# counting letters
def count_letter(text):
    letters = 0
    for char in text:
        if char.isalpha():
            letters += 1
    return letters

# counting sentences
def count_sentence(text):
    sentences = text.count('.')
    sentences += text.count('!')
    sentences += text.count('?')
    return sentences

# caculating grade
def caculate_grade(words, letters, sentences):
    L = letters / words * 100
    S = sentences / words * 100

    grade = 0.0588 * L - 0.296 * S - 15.8

    float_decimal = grade - int(grade)
    return grade + 1 if float_decimal >= 0.5 else int(grade)


# result
words = count_word(text)
letters = count_letter(text)
sentences = count_sentence(text)

grade = caculate_grade(words, letters, sentences)


if grade >= 16:
    print('Grade 16+')
elif grade < 1:
    print('Before Grade 1')
else:
    print(f'Grade {grade}')

File: problem_2/readability/readability.c
#include <cs50.h>
#include <stdio.h>
#include <ctype.h>

int count_letters(string text)
{
    int letters = 0;
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }

    return letters;
}

int count_words(string text)
{
    int words = 0;
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == ' ')
        {
            words++;
        }
    }

    return words + 1;
}

int count_sentences(string text)
{
    int sentences = 0;
    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
    }

    return sentences;
}

int cacul_grade(int letters, int words, int sentences)
{
    double L = (double)letters / words * 100;
    double S = (double)sentences / words * 100;

    double index = 0.0588 * L - 0.296 * S - 15.8;

    // round index
    double the_decimal = index - (int)index;

    return (the_decimal >= 0.5) ? index + 1 : index;
}

int main(void)
{
    string text = get_string("Text: ");

    // counting letter
    int letters = count_letters(text);

    // counting word
    int words = count_words(text);

    // counting sentence
    int sentences = count_sentences(text);

    // caculate grade
    int grade = cacul_grade(letters, words, sentences);

    if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}